File: internal/server/game_manager.go
package server

import "github.com/lemunozm/ascii-arena/internal/server/match"
import "github.com/lemunozm/ascii-arena/internal/pkg/comm"
import "github.com/lemunozm/ascii-arena/internal/pkg/logger"

type GameManager struct {
	playerRegistry *PlayerRegistry
	mapConfig      MapConfig
	startedGame    bool
}

func NewGameManager(maxPlayers int, pointsToWin int, mapConfig MapConfig) *GameManager {
	m := &GameManager{
		playerRegistry: NewPlayerRegistry(maxPlayers, pointsToWin),
		mapConfig:      mapConfig,
		startedGame:    false,
	}

	return m
}

func (s GameManager) GetPlayerRegistry() *PlayerRegistry {
	return s.playerRegistry
}

func (s GameManager) IsGameStarted() bool {
	return s.startedGame
}

func (s GameManager) IsReadyToStartGame() bool {
	return !s.IsGameStarted() && s.playerRegistry.IsFull()
}

func (s *GameManager) RegisterPlayer(connection *comm.Connection) {
	var loginStatus comm.LoginStatus
	if s.IsGameStarted() {
		loginStatus = comm.LOGIN_ERR_GAME_STARTED
		logger.PrintfInfo("Login error: game started")
	} else {
		newPlayerMessage := comm.NewPlayerMessage{}
		connection.Receive(&newPlayerMessage)

		status := s.playerRegistry.Add(newPlayerMessage.Character, connection)

		switch status {
		case REGISTRY_OK:
			loginStatus = comm.LOGIN_SUCCESSFUL
			logger.PrintfInfo("Login succesful (player %c)", newPlayerMessage.Character)
		case REGISTRY_ALREADY_EXISTS:
			loginStatus = comm.LOGIN_ERR_CHARACTER_EXISTS
			logger.PrintfInfo("Login error (player %c): character exists", newPlayerMessage.Character)
		case REGISTRY_FULL: //teorically, it must not happen
			loginStatus = comm.LOGIN_ERR_FULL_PLAYERS
			logger.PrintfInfo("Login error (player %c): full players", newPlayerMessage.Character)
		}
	}

	playerLoginStatusMessage := comm.PlayerLoginStatusMessage{loginStatus}
	connection.Send(playerLoginStatusMessage)

	s.notifyPlayers(connection, loginStatus)
}

func (s *GameManager) notifyPlayers(connection *comm.Connection, loginStatus comm.LoginStatus) {
	playersInfoMessage := comm.PlayersInfoMessage{s.playerRegistry.GetCharacters()}

	if loginStatus == comm.LOGIN_SUCCESSFUL {
		s.sendToAllPlayers(playersInfoMessage)
	} else {
		connection.Send(playersInfoMessage)
	}
}

func (s *GameManager) StartGame() {
	s.startedGame = true
	logger.PrintfInfo("Start game")

	for !s.playerRegistry.HasWinner() {
		s.startMatch()
	}

	logger.PrintfInfo("Finish game")
	s.startedGame = false
}

func (s *GameManager) startMatch() {
	logger.PrintfInfo("Start match")

	arena := match.NewArena(s.mapConfig.Width, s.mapConfig.Height, s.mapConfig.Seed, s.playerRegistry.GetCharacters())

	charactersInfo := make([]comm.CharacterInfo, 0, len(arena.GetCharacters()))
	for _, c := range arena.GetCharacters() {
		charactersInfo = append(charactersInfo, comm.CharacterInfo{c.GetRepresentation(), c.GetPosition()})
	}

	matchInfoMessage := comm.MatchInfoMessage{
		Width:      arena.GetMap().GetWidth(),
		Height:     arena.GetMap().GetHeight(),
		MapSeed:    arena.GetMap().GetSeed(),
		MapData:    arena.GetMap().GetData(),
		Characters: charactersInfo,
	}
	s.sendToAllPlayers(matchInfoMessage)

	for true {
		//TODO
	}

	logger.PrintfInfo("Finish match")
}

func (s *GameManager) sendToAllPlayers(message interface{}) {
	for _, p := range s.playerRegistry.GetPlayers() {
		p.GetConnection().Send(message)
	}
}

File: src/com/asciiarena/lib/server/GameManager.java
package com.asciiarena.lib.server;

import com.asciiarena.lib.common.communication.Connection;
import com.asciiarena.lib.common.communication.ConnectionError;
import com.asciiarena.lib.common.communication.Message;
import com.asciiarena.lib.common.logging.Log;
import com.asciiarena.lib.server.player.PlayerRegistry;

public class GameManager
{
    private final ServerConfig.GameConfig config;
    private PlayerRegistry playerRegistry;
    private MatchManager currentMatch;

    public GameManager(ServerConfig.GameConfig config)
    {
        this.config = config;
        this.playerRegistry = new PlayerRegistry(config.players, config.pointsToWin);
        this.currentMatch = null;
    }

    public boolean checkGameInfo(Connection connection) throws ConnectionError
    {
        Message.GameInfo gameInfoMessage = new Message.GameInfo();
        gameInfoMessage.players = playerRegistry.getCharacters();
        gameInfoMessage.maxPlayers = playerRegistry.getMaxPlayers();
        gameInfoMessage.pointsToWin = playerRegistry.getPointsToWin();
        gameInfoMessage.mapWidth = config.map.width; 
        gameInfoMessage.mapHeight = config.map.height; 
        gameInfoMessage.defaultMapSeed = config.map.seed; 
        connection.send(gameInfoMessage);

        return !playerRegistry.isComplete();
    }

    public boolean login(Connection connection) throws ConnectionError
    {
        Message.NewPlayer newPlayerMessage = (Message.NewPlayer) connection.receive();
        Message.PlayerLogin playerLoginMessage = new Message.PlayerLogin(); 

        synchronized(this)
        {
            playerLoginMessage.status = registerPlayer(newPlayerMessage.character, connection);
            connection.send(playerLoginMessage);

            if(playerLoginMessage.status == Message.PlayerLogin.Status.SUCCESSFUL)
            {
                Message.PlayersInfo playersInfoMessage = new Message.PlayersInfo(playerRegistry.getCharacters()); 
                playerRegistry.sendToPlayers(playersInfoMessage);
            }
        }

        return playerLoginMessage.status == Message.PlayerLogin.Status.SUCCESSFUL;
    }

    private Message.PlayerLogin.Status registerPlayer(char character, Connection connection)
    {
        switch(playerRegistry.add(character, connection))
        {
            case OK:
                Log.info("Login player '%c': successful", character);
                return Message.PlayerLogin.Status.SUCCESSFUL;

            case ALREADY_EXISTS:
                Log.warning("Login player '%c': character already exists", character);
                return Message.PlayerLogin.Status.CHARACTER_ALREADY_EXISTS;

            case FULL:
                Log.warning("Login player '%c': game complete", character);
                return Message.PlayerLogin.Status.GAME_COMPLETE;

            default:
                return Message.PlayerLogin.Status.GAME_COMPLETE;
        }
    }

    public void startGame()
    {
        Log.info("Start game");

        while(!playerRegistry.hasWinner())
        {
            currentMatch = new MatchManager(config.map, playerRegistry); 
            currentMatch.startMatch();
            try
            {
                Thread.sleep(3600000);
            }
            catch (InterruptedException e)
            {
                e.printStackTrace();
            }
            System.exit(1);
        }

        currentMatch = null;

        Log.info("Finish game");
    }

    public void reset()
    {
        playerRegistry = new PlayerRegistry(config.players, config.pointsToWin);
    }

    public boolean isGameStarted()
    {
        return currentMatch != null;
    }

    public MatchManager getCurrentMatch()
    {
        return  currentMatch;
    }

    public PlayerRegistry getPlayerRegistry()
    {
        return playerRegistry;
    }
}
